drivers/hmis/app/models/hmis/hud/enrollment.rb

```diff
@@ -281,6 +281,68 @@ def self.generate_enrollment_id
     generate_uuid
   end
 
+  # Data Collection Features that are enabled for this enrollment, thru its project (e.g. Current Living Situation)
+  #
+  # Is it enabled?
+  #   If ANY data exists for it in this enrollment, even if no active instances exist, then yes.
+  #   (This supports both inactive instances and migrated-in data, as well as context changes such as change in HoH)
+  #
+  # Who is data collected about?
+  #   Choose the "best" instance – IE the one that would actually be selected
+  #   when creating a new record – and return the data_collected_about on it.
+  #
+  # Returns an OpenStruct that is resolved by the DataCollectionFeature GQL type.
+  def data_collection_features
+    # Create OpenStruct for each enabled feature
+    Hmis::Form::Definition::DATA_COLLECTION_FEATURE_ROLES.map do |role|
+      instance_scope = Hmis::Form::Instance.with_role(role).active.published
+      # Service instances must specify a service type or category.
+      instance_scope = instance_scope.for_services if role == :SERVICE
+
+      # Choose the "best" instance, i.e. the one that would actually be selected when recording new data.
+      # We need to do this so that we can accurately set "data collected about" based on the most applicable form.
+      # This detect_best_instance_for_project logic is also called from the query to resolve the actual form definition
+      best_instance = instance_scope.detect_best_instance_for_project(project: project)
+
+      has_any_data = case role
+      when :CURRENT_LIVING_SITUATION
+        current_living_situations.exists?
+      when :SERVICE
+        custom_services.exists? || services.exists?
+      when :CE_EVENT
+        events.exists?
+      when :CE_ASSESSMENT
+        assessments.exists?
+      when :CASE_NOTE
+        custom_case_notes.exists?
+      when :REFERRAL
+        false # Referrals are resolved only on projects, not enrollments
+      when :REFERRAL_REQUEST
+        false # Referrals are resolved only on projects, not enrollments
+      when :EXTERNAL_FORM
+        false # Not resolved on enrollments
+      else
+        raise "Unexpected data collection feature role: #{role}"
+      end
+
+      next unless best_instance || has_any_data
+
+      matches_enrollment = best_instance&.project_and_enrollment_match(project: project, enrollment: self)
+
+      # If an instance applies to this project but doesn't match this enrollment, return, UNLESS there is existing data.
```

Comment:
https://github.com/greenriver/hmis-warehouse/blob/fce7cb27f276f99970c219e555b4567d16b58ca4/drivers/hmis/app/models/hmis/enrollment_assessment_eligibility_list.rb#L77-L87

You can use code similar to this to find the "best instance" while taking into account both the Project and the Enrollment. Find all the published definitions for this role, and rank them by their `project_and_enrollment_match` to find the one that's "best" for this specific Enrollment context. (For services, we only care if _any_ applicable form exists, because none are really "best" as discussed. For non-Service features, we only care which is "best" insofar as it determines what the "data_collected_about" value is.).

It seems like we may not be respecting the DataCollectedAbout value when choosing _which_  form to use for non-assessment data collection (record_form_definition query), which is a separate issue/bug. But that shouldn't stop us from implementing the right logic here.

Reply:
Thanks for pointing out the bug!